docs/rfc/0002_discoveryjob_design.md

```diff
@@ -0,0 +1,196 @@
+- As a user, I want to configure the discovery schedule of a Registry so the "discovery operation" could be triggered automatically by controller.
+
+
+# Detailed design
+
+[design]: #detailed-design
+
+## CRD
+
+The following CRDs will be extended or added to the cluster.
+
+### Registry
+
+Registry represents a registry to be scanned. It contains the registry URL, the name of the secret containing auth credentials, the repositories to be scanned.
+It also contains settings of the discovery and scan schedules.
+
+```yaml
+apiVersion: scanner.rancher.io/v1alpha1
+kind: Registry
+metadata:
+  name: registry-example
+  namespace: default
+  annotations:
+    sbombastic.rancher.io/last-discovery-type:
+    sbombastic.rancher.io/last-discovery-started-at:
+    sbombastic.rancher.io/last-discovered-at
+    sbombastic.rancher.io/last-discoveryjob-name:
+    sbombastic.rancher.io/last-discoveryjob-generation
+    sbombastic.rancher.io/last-discoveryjob-uid
+    sbombastic.rancher.io/running-discovery-type:
+    sbombastic.rancher.io/running-discovery-started-at
+    sbombastic.rancher.io/running-discoveryjob-name:
+    sbombastic.rancher.io/running-discoveryjob-generation
+    sbombastic.rancher.io/running-discoveryjob-uid
+spec:
+  uri: "https://registry-1.docker.io"
+  discoverySchedule:  ## for scheduled discovery
+    cronSchedule: # multiple schedules are supported
+      - "0 1 * * *"
+    discoveryPeriod: "1h" # discovery new images every 1 hour
+  scanSchedule:  ## for scheduled scan (not covered by this doc)
+    scanPeriod: "1d" # scan images every day
+  dbChangedRescan: false
+  repositories: # optional, if not specified, scan all repositories
+    - "repo1"
+    - "repo2"
+```
+
+### DiscoveryJob
+
+A DiscoveryJob represents a discovery operation that can be triggered manually.
+It tracks the status condition of the discovery operation.
+
+```yaml
+apiVersion: scanner.rancher.io/v1alpha1
+kind: DiscoveryJob
+metadata:
+  name: discovery-job-example
+  namespace: namespace-1
+spec:
+  registry: registry-example # registry name
+status:
+  conditions:
+    - type: "Progressing"
+      status: "True"
+      reason: "Finished"
+      ...
+    - type: "Completed"
+      status: "True"
+      reason: "Succeeded"
+      ...
+```
+
+
+1. On Rancher UI the schedule setting should be configured on the Registry page.
+
+2. When a Registry CR is reconciled, a DiscoveryJob CR named(by default) discovery-job-{registry.name} is created for storing the scheduled discovery status.
+   Before a new scheduled discovery operation is triggered, status in this DiscoveryJob CR is purged first
+
+3. Controller caches the scheduled discovery settings of all registries.
+   When a Registry CR is reconciled, controller updates this cache accordingly.
+   There is a scheduler go-routine wakes up per-minute and checks the scheduled discovery settings of all registries.
```

Comment:
This could be a problem if multiple replicas are started. 
We don't want the non-leader replicas to run the scheduler.
Fortunately, this is possible by using the low-level [RunnableFunc](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#RunnableFunc) from controller-runtime.

Reply:
Could we leverage leaderelection.RunOrDie() on a resourcelock.LeaseLock object considering controller has the k8s RBAC for leases?

Reply:
`RunnableFunc` is the proper way to do so, using leader election would not start the func when the manager has started and we don't want to deal with that.
Please take a look to: https://github.com/kubewarden/kubewarden-controller/blob/596315257f952a9572d1384c1fa2cffa9212dcc9/internal/controller/cert_controller.go#L33